pkg/crdutil/crdutil.go

```diff
@@ -84,38 +78,59 @@ func EnsureCRDsCmd() {
 		log.Fatalf("Failed to create API extensions client: %v", err)
 	}
 
-	if err := walkCrdsDir(ctx, client.ApiextensionsV1().CustomResourceDefinitions()); err != nil {
-		log.Fatalf("Failed to apply CRDs: %v", err)
+	dirsToApply, err := walkCRDs(recursive, filenames)
+	if err != nil {
+		log.Fatalf("Failed to walk through CRDs: %v", err)
+	}
+
+	for _, dir := range dirsToApply {
+		log.Printf("Apply CRDs from file: %s", dir)
+		if err := applyCRDs(ctx, client.ApiextensionsV1().CustomResourceDefinitions(), dir); err != nil {
+			log.Fatalf("Failed to apply CRDs: %v", err)
+		}
 	}
 }
 
-// walkCrdsDir walks the CRDs directory and applies each YAML file.
-func walkCrdsDir(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface) error {
-	for _, crdDir := range crdsDir {
+// walkCRDs walks the CRDs directory and applies each YAML file.
+// TODO: add unit test  for this function.
+func walkCRDs(recursive bool, crdDirs []string) ([]string, error) {
+	var dirs []string
+	for _, crdDir := range crdDirs {
+		// We need the parent directory to check if we are in the top-level directory.
+		// This is necessary for the recursive logic.
+		// We can skip the errors as it has been checked in initFlags.
+		parentDir, _ := os.Stat(crdDir)
 		// Walk the directory recursively and apply each YAML file.
 		err := filepath.Walk(crdDir, func(path string, info os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
-			if info.IsDir() || filepath.Ext(path) != ".yaml" {
+			// If this is a directory, skip it.
+			// filepath.Walk() is also called for directories, but we only want to apply CRDs from files.
+			if info.IsDir() {
 				return nil
 			}
-
-			log.Printf("Apply CRDs from file: %s", path)
-			if err := applyCRDsFromFile(ctx, crdClient, path); err != nil {
-				return fmt.Errorf("apply CRD %s: %w", path, err)
+			if filepath.Ext(path) != ".yaml" && filepath.Ext(path) != ".yml" {
+				return nil
 			}
+			// If not recursive we want to only apply the CRDs in the top-level directory.
+			// filepath.Dir() does not add a trailing slash, thus we need to trim it in the crdDir.
+			if !recursive && parentDir.IsDir() && filepath.Dir(path) != strings.TrimRight(crdDir, "/") {
+				return nil
+			}
+
+			dirs = append(dirs, path)
 			return nil
 		})
 		if err != nil {
-			return fmt.Errorf("walk the path %s: %w", crdsDir, err)
+			return []string{}, fmt.Errorf("walk the path %s: %w", crdDirs, err)
 		}
 	}
-	return nil
+	return dirs, nil
 }
 
-// applyCRDsFromFile reads a YAML file, splits it into documents, and applies each CRD to the cluster.
-func applyCRDsFromFile(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface, filePath string) error {
+// applyCRDs reads a YAML file, splits it into documents, and applies each CRD to the cluster.
+func applyCRDs(ctx context.Context, crdClient v1.CustomResourceDefinitionInterface, filePath string) error {
```

Comment:
This rename wasn't strictly required, correct?

Reply:
No, it's not required. It was a review finding https://github.com/NVIDIA/k8s-operator-libs/pull/58#discussion_r1859242624